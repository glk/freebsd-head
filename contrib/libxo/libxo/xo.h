/*
 * Copyright (c) 2014, Juniper Networks, Inc.
 * All rights reserved.
 * This SOFTWARE is licensed under the LICENSE provided in the
 * ../Copyright file. By downloading, installing, copying, or otherwise
 * using the SOFTWARE, you agree to be bound by the terms of that
 * LICENSE.
 * Phil Shafer, July 2014
 */

/**
 * libxo provides a means of generating text, XML, and JSON output
 * using a single set of function calls, maximizing the value of output
 * while minimizing the cost/impact on the code.
 */

#ifndef INCLUDE_XO_H
#define INCLUDE_XO_H

/** Formatting types */
typedef unsigned xo_style_t;
#define XO_STYLE_TEXT	0	/** Generate text output */
#define XO_STYLE_XML	1	/** Generate XML output */
#define XO_STYLE_JSON	2	/** Generate JSON output */
#define XO_STYLE_HTML	3	/** Generate HTML output */

/** Flags for libxo */
typedef unsigned long xo_xof_flags_t;
#define XOF_CLOSE_FP	(1<<0)	/** Close file pointer on xo_close() */
#define XOF_PRETTY	(1<<1)	/** Make 'pretty printed' output */
#define XOF_DIV_OPEN	(1<<2)	/** Internal use only: a <div> is open */
#define XOF_LINE_OPEN	(1<<3)	/** Internal use only: a <div class="line"> */

#define XOF_WARN	(1<<4)	/** Generate warnings for broken calls */
#define XOF_XPATH	(1<<5)	/** Emit XPath attributes in HTML  */
#define XOF_INFO	(1<<6)	/** Emit additional info fields (HTML) */
#define XOF_WARN_XML	(1<<7)	/** Emit warnings in XML (on stdout) */

#define XOF_NO_ENV	(1<<8)	/** Don't look at the LIBXO_OPTIONS env var */
#define XOF_NO_VA_ARG	(1<<9)	/** Don't advance va_list w/ va_arg() */
#define XOF_DTRT	(1<<10)	/** Enable "do the right thing" mode */
#define XOF_KEYS	(1<<11)	/** Flag 'key' fields for xml and json */

#define XOF_IGNORE_CLOSE (1<<12) /** Ignore errors on close tags */
#define XOF_NOT_FIRST	(1<<13)	 /* Not the first item (JSON)  */
#define XOF_NO_LOCALE	(1<<14)	 /** Don't bother with locale */
#define XOF_TOP_EMITTED	(1<<15)	 /* The top JSON braces have been emitted  */

#define XOF_NO_TOP	(1<<16)	/** Don't emit the top braces in JSON */
#define XOF_ANCHOR	(1<<17)	/** An anchor is in place  */
#define XOF_UNITS	(1<<18)	/** Encode units in XML */
#define XOF_UNITS_PENDING (1<<19) /** We have a units-insertion pending */

#define XOF_UNDERSCORES	(1<<20)	/** Replace dashes with underscores (JSON)  */
#define XOF_COLUMNS	(1<<21)	/** xo_emit should return a column count */
#define XOF_FLUSH	(1<<22)	/** Flush after each xo_emit call */

/*
 * The xo_info_t structure provides a mapping between names and
 * additional data emitted via HTML.
 */
typedef struct xo_info_s {
    const char *xi_name;	/* Name of the element */
    const char *xi_type;	/* Type of field */
    const char *xi_help;	/* Description of field */
} xo_info_t;

struct xo_handle_s;		/* Opaque structure forward */
typedef struct xo_handle_s xo_handle_t; /* Handle for XO output */

typedef int (*xo_write_func_t)(void *, const char *);
typedef void (*xo_close_func_t)(void *);
typedef void *(*xo_realloc_func_t)(void *, size_t);
typedef void (*xo_free_func_t)(void *);

/*
 * The formatter function mirrors "vsnprintf", with an additional argument
 * of the xo handle.  The caller should return the number of bytes _needed_
 * to fit the data, even if this exceeds 'len'.
 */
typedef int (*xo_formatter_t)(xo_handle_t *, char *, int,
				const char *, va_list);
typedef void (*xo_checkpointer_t)(xo_handle_t *, va_list, int);

xo_handle_t *
xo_create (xo_style_t style, xo_xof_flags_t flags);

xo_handle_t *
xo_create_to_file (FILE *fp, xo_style_t style, xo_xof_flags_t flags);

void
xo_destroy (xo_handle_t *xop);

void
xo_set_writer (xo_handle_t *xop, void *opaque, xo_write_func_t write_func,
	       xo_close_func_t close_func);

void
xo_set_allocator (xo_realloc_func_t realloc_func, xo_free_func_t free_func);

void
xo_set_style (xo_handle_t *xop, xo_style_t style);

xo_style_t
xo_get_style (xo_handle_t *xop);

int
xo_set_style_name (xo_handle_t *xop, const char *style);

int
xo_set_options (xo_handle_t *xop, const char *input);

xo_xof_flags_t
xo_get_flags (xo_handle_t *xop);

void
xo_set_flags (xo_handle_t *xop, xo_xof_flags_t flags);

void
xo_clear_flags (xo_handle_t *xop, xo_xof_flags_t flags);

void
xo_set_info (xo_handle_t *xop, xo_info_t *infop, int count);

void
xo_set_formatter (xo_handle_t *xop, xo_formatter_t func, xo_checkpointer_t);

void
xo_set_depth (xo_handle_t *xop, int depth);

int
xo_emit_hv (xo_handle_t *xop, const char *fmt, va_list vap);

int
xo_emit_h (xo_handle_t *xop, const char *fmt, ...);

int
xo_emit (const char *fmt, ...);

int
xo_open_container_h (xo_handle_t *xop, const char *name);

int
xo_open_container (const char *name);

int
xo_open_container_hd (xo_handle_t *xop, const char *name);

int
xo_open_container_d (const char *name);

int
xo_close_container_h (xo_handle_t *xop, const char *name);

int
xo_close_container (const char *name);

int
xo_close_container_hd (xo_handle_t *xop);

int
xo_close_container_d (void);

int
xo_open_list_h (xo_handle_t *xop, const char *name);

int
xo_open_list (const char *name);

int
xo_open_list_hd (xo_handle_t *xop, const char *name);

int
xo_open_list_d (const char *name);

int
xo_close_list_h (xo_handle_t *xop, const char *name);

int
xo_close_list (const char *name);

int
xo_close_list_hd (xo_handle_t *xop);

int
xo_close_list_d (void);

int
xo_open_instance_h (xo_handle_t *xop, const char *name);

int
xo_open_instance (const char *name);

int
xo_open_instance_hd (xo_handle_t *xop, const char *name);

int
xo_open_instance_d (const char *name);

int
xo_close_instance_h (xo_handle_t *xop, const char *name);

int
xo_close_instance (const char *name);

int
xo_close_instance_hd (xo_handle_t *xop);

int
xo_close_instance_d (void);

int
xo_attr_h (xo_handle_t *xop, const char *name, const char *fmt, ...);

int
xo_attr_hv (xo_handle_t *xop, const char *name, const char *fmt, va_list vap);

int
xo_attr (const char *name, const char *fmt, ...);

void
xo_error_hv (xo_handle_t *xop, const char *fmt, va_list vap);

void
xo_error_h (xo_handle_t *xop, const char *fmt, ...);

void
xo_error (const char *fmt, ...);

void
xo_flush_h (xo_handle_t *xop);

void
xo_flush (void);

void
xo_finish_h (xo_handle_t *xop);

void
xo_finish (void);

void
xo_set_leading_xpath (xo_handle_t *xop, const char *path);

void
xo_warn_hc (xo_handle_t *xop, int code, const char *fmt, ...);

void
xo_warn_c (int code, const char *fmt, ...);

void
xo_warn (const char *fmt, ...);

void
xo_warnx (const char *fmt, ...);

void
xo_err (int eval, const char *fmt, ...);

void
xo_errx (int eval, const char *fmt, ...);

void
xo_errc (int eval, int code, const char *fmt, ...);

void
xo_message_hcv (xo_handle_t *xop, int code, const char *fmt, va_list vap);

void
xo_message_hc (xo_handle_t *xop, int code, const char *fmt, ...);

void
xo_message_c (int code, const char *fmt, ...);

void
xo_message (const char *fmt, ...);

void
xo_no_setlocale (void);

int
xo_parse_args (int argc, char **argv);

/*
 * This is the "magic" number returned by libxo-supporting commands
 * when passed the equally magic "--libxo-check" option.  If you
 * return this, we can assume that since you know the magic handshake,
 * you'll happily handle future --libxo options and not do something
 * violent like reboot the box or create another hole in the ozone
 * layer.
 */
#define XO_HAS_LIBXO	121

/*
 * externs for our version number strings
 */
extern const char xo_version[];
extern const char xo_version_extra[];

#endif /* INCLUDE_XO_H */
