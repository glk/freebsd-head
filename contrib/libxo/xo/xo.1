.\" #
.\" # Copyright (c) 2014, Juniper Networks, Inc.
.\" # All rights reserved.
.\" # This SOFTWARE is licensed under the LICENSE provided in the
.\" # ../Copyright file. By downloading, installing, copying, or 
.\" # using the SOFTWARE, you agree to be bound by the terms of that
.\" # LICENSE.
.\" # Phil Shafer, July 2014
.\" 
.Dd July, 2014
.Dt LIBXO 3
.Os
.Sh NAME
.Nm xo
.Nd emit formatted output based on format string and arguments
.Sh SYNOPSIS
.Nm xo
.Op Fl options
.Op Ar argument...
.Sh DESCRIPTION
The
.Nm xo
utility allows command line access to the functionality of
the
.Em libxo
library.  Using
.Nm xo ,
shell scripts can emit
.Em XML ,
.Em JSON , or
.Em HTML
using the same commands that emit text output.
.Pp
.Bl -tag -width "12345678901234567"
.It Fl "-close <path>"
Close tags for the given path
.It Fl "-depth <num>"
Set the depth for pretty printing
.It Fl "-help"
Display this help text
.It Fl "-html OR -H"
Generate HTML output
.It Fl "-json OR -J"
Generate JSON output
.It Fl "-leading-xpath <path>"
Add a prefix to generated XPaths (HTML)
.It Fl "-open <path>"
Open tags for the given path
.It Fl "-pretty OR -p"
Make 'pretty' output (add indent, newlines)
.It Fl "-style <style>"
Generate given style (xml, json, text, html)
.It Fl "-text OR -T"
Generate text output (the default style)
.It Fl "-version"
Display version information
.It Fl "-warn OR -W"
Display warnings in text on stderr
.It Fl "-warn-xml"
Display warnings in xml on stdout
.It Fl "-wrap <path>"
Wrap output in a set of containers
.It Fl "-xml OR -X"
Generate XML output
.It Fl "-xpath"
Add XPath data to HTML output);
.El
.Pp
The
.Nm xo
utility accepts a format string suitable for
.Xr xo_emit 3
and a set of zero or more arguments used to supply data for that string.
.Bd -literal -offset indent
    xo "The {k:name} weighs {:weight/%d} pounds.\n" fish 6

  TEXT:
    The fish weighs 6 pounds.
  XML:
    <name>fish</name>
    <weight>6</weight>
  JSON:
    "name": "fish",
    "weight": 6
  HTML:
    <div class="line">
      <div class="text">The </div>
      <div class="data" data-tag="name">fish</div>
      <div class="text"> weighs </div>
      <div class="data" data-tag="weight">6</div>
      <div class="text"> pounds.</div>
    </div>
.Ed
.Pp
The
.Fl "-wrap <path>"
option can be used to wrap emitted content in a
specific hierarchy.  The path is a set of hierarchical names separated
by the '/' character.
.Bd -literal -offset indent
    xo --wrap top/a/b/c '{:tag}' value

  XML:
    <top>
      <a>
        <b>
          <c>
            <tag>value</tag>
          </c>
        </b>
      </a>
    </top>
  JSON:
    "top": {
      "a": {
        "b": {
          "c": {
            "tag": "value"
          }
        }
      }
    }
.Ed
.Pp
The
.Fl "\-open <path>"
and
.Fl "\-close <path>"
can be used to emit
hierarchical information without the matching close and open
tag.  This allows a shell script to emit open tags, data, and
then close tags.  The
.Fl \-depth
option may be used to set the
depth for indentation.  The
.Fl "\-leading-xpath"
may be used to
prepend data to the XPath values used for HTML output style.
.Bd -literal -offset indent
    #!/bin/sh
    xo --open top/data
    xo --depth 2 '{tag}' value
    xo --close top/data
  XML:
    <top>
      <data>
        <tag>value</tag>
      </data>
    </top>
  JSON:
    "top": {
      "data": {
        "tag": "value"
      }
    }
.Ed
.Pp
.Sh EXAMPLE
.Bd -literal -offset indent
  % xo 'The {:product} is {:status}\n' stereo "in route"
  The stereo is in route
  % xo -p -X 'The {:product} is {:status}\n' stereo "in route"
  <product>stereo</product>
  <status>in route</status>
.Ed
.Pp
.Sh ADDITIONAL DOCUMENTATION
.Pp
Complete documentation can be found on github:
.Bd -literal -offset indent
http://juniper.github.io/libxo/libxo-manual.html
.Ed
.Pp
libxo lives on github as:
.Bd -literal -offset indent
https://github.com/Juniper/libxo
.Ed
.Pp
The latest release of libxo is available at:
.Bd -literal -offset indent
https://github.com/Juniper/libxo/releases
.Ed
.Sh SEE ALSO
.Xr xo_emit 3
.Sh HISTORY
The
.Fa libxo
library was added in FreeBSD 11.0.
.Sh AUTHOR
Phil Shafer
