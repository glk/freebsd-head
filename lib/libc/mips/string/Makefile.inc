#	$NetBSD: Makefile.inc,v 1.2 2000/10/10 21:51:54 jeffs Exp $
# $FreeBSD$

SRCS+=	bcmp.S bcopy.S bzero.S ffs.S memchr.c memcmp.c memset.c \
	memcpy.S memmove.S \
	strcat.c strchr.S strcmp.S strcpy.c strcspn.c strlen.S \
	strncat.c strncmp.c strncpy.c strrchr.S strpbrk.c strsep.c \
	strspn.c strstr.c swab.c
