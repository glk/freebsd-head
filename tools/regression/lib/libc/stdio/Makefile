# $FreeBSD$

TESTS=	test-fdopen \
	test-fmemopen \
	test-fopen \
	test-freopen \
	test-getdelim \
	test-mkostemp \
	test-open_memstream \
	test-open_wmemstream \
	test-perror \
	test-print-positional \
	test-printbasic \
	test-printfloat \
	test-scanfloat 
CFLAGS+= -lm

.PHONY: tests
tests: ${TESTS}
	for p in ${TESTS}; do ${.OBJDIR}/$$p; done

.PHONY: clean
clean:
	-rm -f ${TESTS}
