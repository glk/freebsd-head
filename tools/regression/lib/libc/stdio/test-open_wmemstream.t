#!/bin/sh
# $FreeBSD$

cd `dirname $0`

executable=`basename $0 .t`

make $executable 2>&1 > /dev/null

echo 1..1
if ./$executable; then
	echo ok 1 - $executable successful
else
	echo not ok 1 - $executable failed
fi
