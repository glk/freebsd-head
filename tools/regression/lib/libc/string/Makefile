# $FreeBSD$

TESTS=	test-stpncpy test-strerror test-wcscasecmp test-wcsnlen

.PHONY: tests
tests: ${TESTS}
	for p in ${TESTS}; do ${.OBJDIR}/$$p; done

.PHONY: clean
clean:
	-rm -f ${TESTS}
