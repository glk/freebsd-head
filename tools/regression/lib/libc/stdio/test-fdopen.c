/*-
 * Copyright (c) 2014 Jilles Tjoelker
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include	<fcntl.h>
#include	<stdbool.h>
#include	<stdio.h>
#include	<string.h>
#include	<unistd.h>

static int testnum = 1;

static void
runtest(const char *fname, int intmode, const char *strmode, bool success)
{
	FILE *fp;
	int fd;

	fd = open(fname, intmode);
	if (fd == -1) {
		printf("not ok %d - open(\"%s\", %#x) failed\n",
		    testnum++, fname, intmode);
		return;
	}
	fp = fdopen(fd, strmode);
	if (fp == NULL) {
		close(fd);
		if (success)
			printf("not ok %d - "
			    "fdopen(open(\"%s\", %#x), \"%s\") failed\n",
			    testnum++, fname, intmode, strmode);
		else
			printf("ok %d - "
			    "fdopen(open(\"%s\", %#x), \"%s\") failed\n",
			    testnum++, fname, intmode, strmode);
		return;
	}
	if (success)
		printf("ok %d - "
		    "fdopen(open(\"%s\", %#x), \"%s\") succeeded\n",
		    testnum++, fname, intmode, strmode);
	else
		printf("not ok %d - "
		    "fdopen(open(\"%s\", %#x), \"%s\") succeeded\n",
		    testnum++, fname, intmode, strmode);
	fclose(fp);
}

/*
 * Test program for fdopen().
 */
int
main(int argc, char *argv[])
{
	printf("1..19\n");
	runtest("/dev/null", O_RDONLY, "r", true);
	runtest("/dev/null", O_WRONLY, "r", false);
	runtest("/dev/null", O_RDWR, "r", true);
	runtest("/dev/null", O_RDONLY, "w", false);
	runtest("/dev/null", O_WRONLY, "w", true);
	runtest("/dev/null", O_RDWR, "w", true);
	runtest("/dev/null", O_RDONLY, "a", false);
	runtest("/dev/null", O_WRONLY, "a", true);
	runtest("/dev/null", O_RDWR, "a", true);
	runtest("/dev/null", O_RDONLY, "r+", false);
	runtest("/dev/null", O_WRONLY, "r+", false);
	runtest("/dev/null", O_RDWR, "r+", true);
	runtest("/dev/null", O_RDONLY, "w+", false);
	runtest("/dev/null", O_WRONLY, "w+", false);
	runtest("/dev/null", O_RDWR, "w+", true);
	runtest("/bin/sh", O_EXEC, "r", false);
	runtest("/bin/sh", O_EXEC, "w", false);
	runtest("/bin/sh", O_EXEC, "r+", false);
	runtest("/bin/sh", O_EXEC, "w+", false);

	return 0;
}

/* vim:ts=8:cin:sw=8
 *  */
