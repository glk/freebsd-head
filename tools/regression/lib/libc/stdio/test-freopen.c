/*-
 * Copyright (c) 2014 Jilles Tjoelker
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include	<stdbool.h>
#include	<stdio.h>
#include	<string.h>

static int testnum = 1;

static void
runtest(const char *fname1, const char *mode1, const char *fname2,
    const char *mode2, bool success)
{
	FILE *fp1, *fp2;
	const char *fname2_print;

	fname2_print = fname2 != NULL ? fname2 : "<NULL>";
	fp1 = fopen(fname1, mode1);
	if (fp1 == NULL) {
		printf("not ok %d - fopen(\"%s\", \"%s\") failed\n",
		    testnum++, fname1, mode1);
		return;
	}
	fp2 = freopen(fname2, mode2, fp1);
	if (fp2 == NULL) {
		fclose(fp1);
		if (success)
			printf("not ok %d - "
			    "freopen(\"%s\", \"%s\", fopen(\"%s\", \"%s\")) "
			    "failed\n",
			    testnum++, fname2_print, mode2, fname1, mode1);
		else
			printf("ok %d - "
			    "freopen(\"%s\", \"%s\", fopen(\"%s\", \"%s\")) "
			    "failed\n",
			    testnum++, fname2_print, mode2, fname1, mode1);
		return;
	}
	if (success)
		printf("ok %d - "
		    "freopen(\"%s\", \"%s\", fopen(\"%s\", \"%s\")) "
		    "succeeded\n",
		    testnum++, fname2_print, mode2, fname1, mode1);
	else
		printf("not ok %d - "
		    "freopen(\"%s\", \"%s\", fopen(\"%s\", \"%s\")) "
		    "succeeded\n",
		    testnum++, fname2_print, mode2, fname1, mode1);
	fclose(fp2);
}

/*
 * Test program for freopen().
 */
int
main(int argc, char *argv[])
{
	printf("1..19\n");
	runtest("/dev/null", "r", NULL, "r", true);
	runtest("/dev/null", "w", NULL, "r", false);
	runtest("/dev/null", "r+", NULL, "r", true);
	runtest("/dev/null", "r", NULL, "w", false);
	runtest("/dev/null", "w", NULL, "w", true);
	runtest("/dev/null", "r+", NULL, "w", true);
	runtest("/dev/null", "r", NULL, "a", false);
	runtest("/dev/null", "w", NULL, "a", true);
	runtest("/dev/null", "r+", NULL, "a", true);
	runtest("/dev/null", "r", NULL, "r+", false);
	runtest("/dev/null", "w", NULL, "r+", false);
	runtest("/dev/null", "r+", NULL, "r+", true);
	runtest("/dev/null", "r", NULL, "w+", false);
	runtest("/dev/null", "w", NULL, "w+", false);
	runtest("/dev/null", "r+", NULL, "w+", true);
	runtest("/bin/sh", "r", NULL, "r", true);
	runtest("/bin/sh", "r", "/bin/sh", "r", true);
	runtest("/bin/sh", "r", "/dev/null", "r", true);
	runtest("/bin/sh", "r", "/dev/null", "w", true);

	return 0;
}

/* vim:ts=8:cin:sw=8
 *  */
