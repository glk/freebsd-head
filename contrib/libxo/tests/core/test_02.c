/*
 * Copyright (c) 2014, Juniper Networks, Inc.
 * All rights reserved.
 * This SOFTWARE is licensed under the LICENSE provided in the
 * ../Copyright file. By downloading, installing, copying, or otherwise
 * using the SOFTWARE, you agree to be bound by the terms of that
 * LICENSE.
 * Phil Shafer, July 2014
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

#include "xo.h"

int
main (int argc, char **argv)
{
    argc = xo_parse_args(argc, argv);
    if (argc < 0)
	return 1;

    for (argc = 1; argv[argc]; argc++) {
	if (strcmp(argv[argc], "xml") == 0)
	    xo_set_style(NULL, XO_STYLE_XML);
	else if (strcmp(argv[argc], "json") == 0)
	    xo_set_style(NULL, XO_STYLE_JSON);
	else if (strcmp(argv[argc], "text") == 0)
	    xo_set_style(NULL, XO_STYLE_TEXT);
	else if (strcmp(argv[argc], "html") == 0)
	    xo_set_style(NULL, XO_STYLE_HTML);
	else if (strcmp(argv[argc], "pretty") == 0)
	    xo_set_flags(NULL, XOF_PRETTY);
	else if (strcmp(argv[argc], "xpath") == 0)
	    xo_set_flags(NULL, XOF_XPATH);
	else if (strcmp(argv[argc], "info") == 0)
	    xo_set_flags(NULL, XOF_INFO);
    }

    xo_set_flags(NULL, XOF_UNITS); /* Always test w/ this */

    xo_open_container_h(NULL, "top");

    xo_open_container("data");

    xo_emit("{:mbuf-current/%u}/{:mbuf-cache/%u}/{:mbuf-total/%u} "
	    "{N:mbufs <&> in use (current\\/cache\\/total)}\n",
	    10, 20, 30);

    xo_emit("{:distance/%u}{Uw:miles} from {:location}\n", 50, "Boston");
    xo_emit("{:memory/%u}{U:k} left out of {:total/%u}{U:kb}\n", 64, 640);
    xo_emit("{:memory/%u}{U:/%s} left out of {:total/%u}{U:/%s}\n",
	    64, "k", 640, "kilobytes");

    xo_emit("{T:/before%safter:}\n", "working");

    xo_emit("{d:some/%s}{:ten/%ju}{:eleven/%ju}\n",
	    "string", (uintmax_t) 10, (uintmax_t) 11);

    xo_emit("{:unknown/%u} "
	    "{N:/packet%s here\\/there\\/everywhere}\n",
	    1010, "s");

    xo_emit("({[:/%d}{n:min/15}/{n:cur/20}/{:max/%d}{]:})\n", 30, 125);
    xo_emit("({[:30}{:min/%u}/{:cur/%u}/{:max/%u}{]:})\n", 15, 20, 125);
    xo_emit("({[:-30}{n:min/15}/{n:cur/20}/{n:max/125}{]:})\n");
    xo_emit("({[:}{:min/%u}/{:cur/%u}/{:max/%u}{]:/%d})\n", 15, 20, 125, -30);

    xo_open_list("flag");
    xo_emit("{lq:flag/one} {lq:flag/two} {lq:flag/three}\n");
    xo_close_list("flag");

    xo_emit("{e:empty-tag/}");
    xo_emit("1:{qt:t1/%*d} 2:{qt:t2/test%-*u} 3:{qt:t3/%10sx} 4:{qt:t4/x%-*.*s}\n",
	    6, 1000, 8, 5000, "ten-long", 10, 10, "test");
    xo_emit("{E:this is an error}\n");
    xo_emit("{E:/%s more error%s}\n", "two", "s" );
    xo_emit("{W:this is an warning}\n");
    xo_emit("{W:/%s more warning%s}\n", "two", "s" );
    xo_emit("{L:/V1\\/V2 packet%s}: {:count/%u}\n", "s", 10);

    int test = 4;
    xo_emit("{:test/%04d} {L:/tr%s}\n", test, (test == 1) ? "y" : "ies");

    xo_message("improper use of profanity; %s; %s",
	       "ten yard penalty", "first down");

    xo_error("Shut 'er down, Clancey!  She's a-pumpin' mud!  <>!,\"!<>\n");

    xo_close_container("data");

    xo_close_container_h(NULL, "top");

    xo_finish();

    return 0;
}
