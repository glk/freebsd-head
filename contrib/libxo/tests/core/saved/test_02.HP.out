<div class="line">
  <div class="data" data-tag="mbuf-current">10</div>
  <div class="text">/</div>
  <div class="data" data-tag="mbuf-cache">20</div>
  <div class="text">/</div>
  <div class="data" data-tag="mbuf-total">30</div>
  <div class="text"> </div>
  <div class="note">mbufs &lt;&amp;&gt; in use (current/cache/total)</div>
</div>
<div class="line">
  <div class="data" data-tag="distance" data-units="miles">50</div>
  <div class="padding"> </div>
  <div class="text"> from </div>
  <div class="data" data-tag="location">Boston</div>
</div>
<div class="line">
  <div class="data" data-tag="memory" data-units="k">64</div>
  <div class="text"> left out of </div>
  <div class="data" data-tag="total" data-units="kb">640</div>
</div>
<div class="line">
  <div class="data" data-tag="memory" data-units="k">64</div>
  <div class="text"> left out of </div>
  <div class="data" data-tag="total" data-units="kilobytes">640</div>
</div>
<div class="line">
  <div class="title">beforeworkingafter:</div>
</div>
<div class="line">
  <div class="data" data-tag="some">string</div>
  <div class="data" data-tag="ten">10</div>
  <div class="data" data-tag="eleven">11</div>
</div>
<div class="line">
  <div class="data" data-tag="unknown">1010</div>
  <div class="text"> </div>
  <div class="note">packets here/there/everywhere</div>
</div>
<div class="line">
  <div class="text">(</div>
  <div class="padding">                     </div>
  <div class="data" data-tag="min">15</div>
  <div class="text">/</div>
  <div class="data" data-tag="cur">20</div>
  <div class="text">/</div>
  <div class="data" data-tag="max">125</div>
  <div class="text">)</div>
</div>
<div class="line">
  <div class="text">(</div>
  <div class="padding">                     </div>
  <div class="data" data-tag="min">15</div>
  <div class="text">/</div>
  <div class="data" data-tag="cur">20</div>
  <div class="text">/</div>
  <div class="data" data-tag="max">125</div>
  <div class="text">)</div>
</div>
<div class="line">
  <div class="text">(</div>
  <div class="data" data-tag="min">15</div>
  <div class="text">/</div>
  <div class="data" data-tag="cur">20</div>
  <div class="text">/</div>
  <div class="data" data-tag="max">125</div>
  <div class="padding">                     </div>
  <div class="text">)</div>
</div>
<div class="line">
  <div class="text">(</div>
  <div class="data" data-tag="min">15</div>
  <div class="text">/</div>
  <div class="data" data-tag="cur">20</div>
  <div class="text">/</div>
  <div class="data" data-tag="max">125</div>
  <div class="padding">                     </div>
  <div class="text">)</div>
</div>
<div class="line">
  <div class="data" data-tag="flag">one</div>
  <div class="text"> </div>
  <div class="data" data-tag="flag">two</div>
  <div class="text"> </div>
  <div class="data" data-tag="flag">three</div>
</div>
<div class="line">
  <div class="text">1:</div>
  <div class="data" data-tag="t1">  1000</div>
  <div class="text"> 2:</div>
  <div class="data" data-tag="t2">test5000    </div>
  <div class="text"> 3:</div>
  <div class="data" data-tag="t3">  ten-longx</div>
  <div class="text"> 4:</div>
  <div class="data" data-tag="t4">xtest      </div>
</div>
<div class="line">
  <div class="error">this is an error</div>
</div>
<div class="line">
  <div class="error">two more errors</div>
</div>
<div class="line">
  <div class="warning">this is an warning</div>
</div>
<div class="line">
  <div class="warning">two more warnings</div>
</div>
<div class="line">
  <div class="label">V1/V2 packets</div>
  <div class="text">: </div>
  <div class="data" data-tag="count">10</div>
</div>
<div class="line">
  <div class="data" data-tag="test">0004</div>
  <div class="text"> </div>
  <div class="label">tries</div>
</div>
<div class="line">
  <div class="message">improper use of profanity; ten yard penalty; first down
</div>
</div>
<div class="line">
  <div class="error">Shut 'er down, Clancey!  She's a-pumpin' mud!  &lt;&gt;!,"!&lt;&gt;
</div>
</div>
